packages/network-controller/src/NetworkController.ts

```diff
@@ -604,6 +665,8 @@ export class NetworkController extends BaseControllerV2<
         ? NetworksTicker[type]
         : 'ETH';
 
+    this.#ensureAutoManagedNetworkClientRegistryPopulated();
```

Comment:
All of the methods that end up setting the network client need to now access the network client registry. This means that at some point, the registry needs to be initialized.

We could initialize the registry in the constructor, but because the registry is reliant on state (`networkConfigurations` and `providerConfig`), if any part of that state is invalid such that a network client cannot be created, the constructor would throw. This would be at odds with how the network controller works presently, as we don't involve the network client until it's set.

Separately, one thing I noticed in the tests is that there are many places where we don't initialize the provider before switching the network. While this is unrealistic, as in the extension or mobile app we always call `initializeProvider` after making a new instance of NetworkController, it doesn't cause a problem.

Therefore, in order to keep the existing behavior, I opted to make the registry initialization lazy. That is, I added code to `initializeProvider`, `setProviderType`, `setActiveNetwork`, `rollbackToPreviousProvider`, and `resetConnection` which will check to see if the registry has been populated and does so if it is not. This effectively solves both of the above constraints. I've also added the same check to `upsertNetworkConfiguration` and `removeNetworkConfiguration`, as neither require the provider to be initialized in order to update the registry.

Reply:
>Separately, one thing I noticed in the tests is that there are many places where we don't initialize the provider before switching the network. While this is unrealistic, as in the extension or mobile app we always call initializeProvider after making a new instance of NetworkController, it doesn't cause a problem.

We might stop initializing the network controller on construction soon. So the effort you've put in here may soon be quite important.

Reply:
No, what I'm saying is that if we constructed the network clients in the constructor, if that throws due to invalid state, it would be different behavior than what happens now. Currently if you pass an invalid provider or network config, the controller is happy to accept it — until you try to set the provider.

Reply:
Right, I understand that this was done to avoid a change in behavior. Just that we don't usually consider what effects invalid input has, and we don't make a point to avoid making functional changes that only affect invalid input. e.g. we wouldn't go out of our way to avoid calling `toLowerCase()` with a string in state just to avoid introducing an error if it was a number. We make changes that would break for invalid all the time, without giving it a second thought.

Reply:
But either way, I'm fine with leaving it and dealing with it later as part of the "selected network validation" issue that I will be creating shortly.

Reply:
Validation issue is here: https://github.com/MetaMask/core/issues/1461

And the related state normalization issue is https://github.com/MetaMask/core/issues/1462